tortoise/tracer.go

```diff
@@ -0,0 +1,487 @@
+package tortoise
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/common/types/result"
+)
+
+type output struct {
+	Type  eventType       `json:"t"`
+	Event json.RawMessage `json:"o"`
+}
+
+type tracer struct {
+	logger *zap.Logger
+}
+
+func (t *tracer) On(event traceEvent) {
+	buf, err := json.Marshal(event)
+	if err != nil {
+		panic(err.Error())
+	}
+	raw := json.RawMessage(buf)
+	t.logger.Info("",
+		zap.Uint16("t", event.Type()),
+		zap.Any("o", &raw),
+	)
+}
+
+type TraceOpt func(*zap.Config)
+
+func WithOutput(path string) TraceOpt {
+	return func(cfg *zap.Config) {
+		cfg.OutputPaths = []string{path}
+	}
+}
+
+func newTracer(opts ...TraceOpt) *tracer {
+	cfg := zap.NewProductionConfig()
+	cfg.Sampling = nil
+	cfg.EncoderConfig.CallerKey = zapcore.OmitKey
+	cfg.EncoderConfig.MessageKey = zapcore.OmitKey
+	cfg.EncoderConfig.LevelKey = zapcore.OmitKey
+	cfg.DisableCaller = true
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	logger, err := cfg.Build()
+	if err != nil {
+		panic(err.Error())
+	}
+	return &tracer{
+		logger: logger.Named("tracer"),
+	}
+}
+
+type traceRunner struct {
+	opts          []Opt
+	trt           *Tortoise
+	pending       map[types.BallotID]*DecodedBallot
+	assertOutputs bool
+	assertErrors  bool
+}
+
+func RunTrace(path string, breakpoint func(), opts ...Opt) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	dec := json.NewDecoder(bufio.NewReaderSize(f, 1<<20))
+	enum := newEventEnum()
+	runner := &traceRunner{
+		opts:          opts,
+		pending:       map[types.BallotID]*DecodedBallot{},
+		assertOutputs: true,
+		assertErrors:  true,
+	}
+	for {
+		ev, err := enum.Decode(dec)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return err
+		}
+		if err := ev.Run(runner); err != nil {
+			return err
+		}
+		if breakpoint != nil {
```

Comment:
Having the breakpoint here means that you can't manipulate the event contents. It seems like that would be something that could be useful in interactive debugging.

Reply:
added breakpoint after ev is assigned

Reply:
sorry, had to revert it. i creates some issues with access to runner. don't want to debug it now
> (dlv) print runner.trt.trtl
(unreadable could not find loclist entry at 0x37045 for address 0xdea7e